### PR TITLE
Clear bot memory appropriately on state transitions

### PR DESCRIPTION
**Pull request by slipher**

### CODE REVIEW

**Review by DolceTriade — COMMENTED** [review]
Did anything depend on the old behavior?

**Comment by sweet235** [comment]
Which bugs exactly?

**Comment by slipher** [comment]
> Which bugs exactly?

This one is the biggest: `Resetting myTimer fixes a bug pointed out by Sweet:
if a bot switches to our default behavior tree from any other one, it
instantly commits suicide (rather than waiting 20 seconds as intended).`

**Comment by sweet235** [comment]
> This one is the biggest

That is a bug in the behaviors `defend.bt`, `stay-here.bt`, `follow.bt`. I simply forgot to update these.

**Comment by slipher** [comment]
I would definitely consider that a bug in the framework more than a bug in the individual BTs. It's absurd to require every BT that uses a timer to have to start with "cleanup" in case it was used by other BTs.

If you want to find out about more bugs see the commit messages. For example userSpecifiedClientNum being set for a bot that just connected.

**Comment by sweet235** [comment]
> Ensure userSpecifiedClientNum and userSpecifiedLocation are cleared if a bot connects in some client slot a second time.

Oops.

> It's absurd to require every BT that uses a timer to have to start with "cleanup" in case it was used by other BTs.

Yes and no. The reason a reset in the framework works is: all our behavior files use this timer for the same thing. A user might use it for something different entirely, and then a reset might not be desired. I agree that _if_ we reset in the framework, the proposed method is the best we can do.

I agree that resetting the bot states is a good thing. But I am not yet convinced about some fields. Should we reset the planned path completely? Or should we make the bot move to the next path corner, and reset then? This might not seem important, but keep in mind: when we reset the navcon informations, all currently climbing bots will fall from the walls instantly. They might fall into a death zone.

**Comment by sweet235** [comment]
> Oops.

Well, that is not elegant, but not a bug. A new bot will be set to use `default.bt` which does not use these values.

**Comment by slipher** [comment]
> I agree that resetting the bot states is a good thing. But I am not yet convinced about some fields. Should we reset the planned path completely? Or should we make the bot move to the next path corner, and reset then? This might not seem important, but keep in mind: when we reset the navcon informations, all currently climbing bots will fall from the walls instantly. They might fall into a death zone.

Hmm, can that really work? I figured after a behavior change the bot would almost always select a new destination and obliterate the old path information. I might do some tests to see if this is the case.



> Well, that is not elegant, but not a bug. A new bot will be set to use `default.bt` which does not use these values.

A `bot add` command may specify a behavior.

**Comment by sweet235** [comment]
> I figured after a behavior change the bot would almost always select a new destination and obliterate the old path information

Sometimes, not almost always. It depends on the behavior, and on the map's geometry and size. If a bot is alerted to an enemy, all existing behaviors will make it go there (except when building or extinguishing fire). On maps like yocto or spacetracks, this is almost always the case.

It is true that currently, a behavior change can make a bot fall from a wall. But with the proposed changes, this will happen always.

**Comment by sweet235** [comment]
> A bot add command may specify a behavior.

The `bot` command is usually restricted to admins. That is a bug indeed, but not a severe bug.

We are planning to allow allow everyone to select behaviors with the `tactic` command. The only reason we do not enable this by default is the lack of GUI. I had hoped than someone would make GUI for that, but that did not happen in the last year. Once we enable this, any oddities at behavior change will be very visible to all users.

**Comment by sweet235** [comment]
>> `memory.currentNode = nullptr;`

> Why is this needed?

You did not answer my question. So I had to investigate this myself. For sequence nodes, the behavior tree evaluator will resume each execution at the currently running node(s). It does that by maintaining a vector of currently running node addresses for each bot.

When changing a bot's behavior, a new tree with physically different nodes is used. For instance: `action roam` in the new tree has a different address than `action roam` in the old tree. So resetting to zero has no observable effect.

But that is not true in all cases. If both behavior tree files have `behavior third` somewhere in the tree, the nodes in `third.bt` will be physically equal in both cases. So the evaluator can resume at nodes in different files. Typical candidates that might be resumed are `action fight`, `action heal`, `action roam`.

Conclusions:

- We should not attempt to reset random fields in the bot minds until we know exactly what the consequences are.
- We might exploit or remove the ability to resume nodes from different files.

File: src/sgame/sg_bot.cpp [hunk 1]
```diff
@@ -798,6 +801,18 @@ void botMemory_t::doSprint( int jumpCost, int stamina, usercmd_t& cmd )
 	exhausted = exhausted && stamina <= jumpCost * 2;
 }
 
+// TODO: also reset state stored in BT nodes
+void G_Bot_ResetBehaviorState( botMemory_t &memory )
+{
+	memory.currentNode = nullptr;
```
[the remark this answers is not in the document]
Comment (slipher) [reply src/sgame/sg_bot.cpp:807]: To terminate any running action. I didn't add it btw, it was already there.

**Comment by slipher** [comment]
TL;DR I removed code to clear `lastNavcon*` on behavior change.

To clarify things, `currentNode`, `runningNodes`, `goal`, and `m_nav` are all cleared on behavior change on master. The original version of this PR only cleared 3 additional navigation-related fields on behavior change:
- `hasOffmeshGoal`
- `lastNavconTime`
- `lastNavconDistance`

`hasOffmeshGoal` does not relate to navcons, but rather targeting things in line of sight without using pathfinding. It is always set in conjunction with a goal change. I'm confident it should be cleared.

The other two are more complicated. It would be nice to test them, but that seemed too hard since I don't know of any publicly available navcon sets including vertical navcons. From reading the code, it seems that they might still be used if the bot's new target happens to have a greater z-coordinate than its current position. Continuing to follow a navcon might "work" by accident in some cases. So I left them as-is for now. There is a lot of other nav state in `Bot_t` that is not yet properly cleared anyway.

By the way I think I found a bug: #3002

**Comment by sweet235** [comment]
> To clarify things, currentNode, runningNodes, goal, and m_nav are all cleared on behavior change on master.

I'm just brainstorming. Could we really make the bot go the next path corner? Maybe. In that case, we might consider not clearing `currentNode`, `runningNodes` and the path.

> I'm confident it should be cleared.

I agree.

> It would be nice to test them, but that seemed too hard since I don't know of any publicly available navcon sets including vertical navcons

A sad truth. If you can tolerate toy maps, this one comes with vertical navcons for all classes: [map-voxelmini0](https://users.unvanquished.net/~sweet/pkg/map-voxelmini0_0.dpk)

**Comment by slipher** [comment]
> I'm just brainstorming. Could we really make the bot go the next path corner? Maybe.

Seems doable. I think the way to do this would be to keep the path fixed while we think we're in a navcon. Let the goal change freely, but don't set a new path until later. Presumably we would want to work not only with behavior changes, but also the more common case of aborting an action before completed. Like when the bot is going one way but gets shot, decides to heal, and turns around.

**Comment by sweet235** [comment]
LGTM
